models/sudoswap/ethereum/sudoswap_ethereum_events.sql

```diff
@@ -111,70 +132,88 @@ WITH
             FROM (
                 SELECT 
                     swaps.*
                     , COALESCE(fu.newfee, pc.initialfee)/1e18 as ownerfee --most recent ownerfee, depends on bonding curve to implement it correctly. See explanation in fee table schema.
                     , COALESCE(fu.evt_block_time, pc.block_time) as fee_update_time
                     , pfu.newMultiplier/1e18 as protocolfee --most recent protocolfee, depends on bonding curve to implement it correctly. See explanation in fee table schema.
-                    , pfu.evt_block_time as protocolfee_updatetime
+                    , pfu.evt_block_time as protocolfee_update_time
+                    , pc.protocolfee_recipient
                     , pc.nftcontractaddress
-                    , pc.asset_recip
-                FROM swaps
+                    , coalesce(aru.a, pc.asset_recip) as asset_recip
+                    , coalesce(aru.evt_block_time, pc.block_time) as asset_recip_update_time
+                FROM swaps_with_calldata swaps
                 JOIN pairs_created pc ON pc.pair_address = contract_address --remember swaps from other NFT addresses won't appear!
+                -- we might need to do these joins separately since we're exploding into a lot of rows..
+                -- should not matter a lot since # of changes per pool should be small
                 LEFT JOIN owner_fee_update fu ON swaps.call_block_time >= fu.evt_block_time AND swaps.contract_address = fu.contract_address
                 LEFT JOIN protocol_fee_update pfu ON swaps.call_block_time >= pfu.evt_block_time
+                LEFT JOIN asset_recipient_update aru on swaps.call_block_time >= aru.evt_block_time AND swaps.contract_address = aru.contract_address
             ) a
         ) b
         WHERE ordering = 1 --we want to keep the most recent ownerfee and protocol fee for each individual call (trade)
-    ),
+    )
 
-    swaps_w_traces as (
+    ,swaps_w_traces as (
         -- we traces to get NFT and ETH transfer data because sudoswap doesn't emit any data in events for swaps, so we have to piece it together manually based on trace_address.
         SELECT
             sb.call_block_time
             , sb.call_block_number
             , sb.trade_category
-            , CASE WHEN sb.trade_category = 'Buy' THEN SUM(value)/(1+sb.ownerfee+sb.protocolfee)
-                ELSE SUM(value)/(1-sb.ownerfee-sb.protocolfee)
-                END as base_price
-            , SUM(value) as trade_price --should give total value of the trade (buy or sell)
+            , SUM(
+                CASE WHEN sb.trade_category = 'Buy' -- caller buys, AMM sells
+                THEN (
+                    CASE WHEN tr.from = sb.call_from THEN value -- amount of ETH payed
+                    WHEN (tr.to = sb.call_from AND sb.call_from != sb.asset_recip) THEN -value --refunds unless the caller is also the asset recipient, no way to discriminate there.
+                    ELSE 0 END)
```

Comment:
I still don't fully understand refunds, mind adding in the comments here an example tx or flow of ETH? is it price paid then partial value is sent back or something?

Reply:
Yes so since these function calls are just 'payable' one can send however much ETH you'd like when interacting with a pool directly. The ETH that was not used for the transfer is sent back to the caller at the end. 
Nice example:
https://etherscan.io/tx/0x1c74e5f2717bc4f9d791ad89183eb351919fdfb7569968a65cddc9a1a7b1b5eb
- 1ETH call `swapTokenForAnyNFTs`
- 0.5010 ETH to the pool recipient
- 0.0025 ETH to the protocol
- 0.4964 ETH refund